docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,246 @@
+# Supporting build artifact dependencies
+
+## Background
+Refer [this](https://tinyurl.com/skaffold-modules) document which presents the design goals for introducing the concept of _modules_ in Skaffold. A prerequisite to being able to define modules and supporting cross module dependency is to first consider the latest version of the skaffold config (_currently `v2beta8`_) as an implicit module and allowing dependencies between artifacts defined in it. The current document aims to capture the major code changes necessary for achieving this.
+
+>*Note: Omitted some details for brevity, assuming reader's familiarity with the skaffold [codebase](https://github.com/GoogleContainerTools/skaffold).*
+
+## Scope
+
+We scope our initial implementation to support artifact dependencies only for docker artifacts. This means that docker artifacts can define dependencies on any other artifact types. It doesn't limit another artifact type from being used as a dependency.
+## Config schema
+
+We introduce an `ArtifactDependency` slice within `Artifact` in `config.go`
```

Comment:
With the new scope, can this part of just the `DockerArtifact` struct? If we don't plan to support modules for other builders in near future, it will make sense to just add it to `DockerArtifact`

Reply:
I want to leave this inside `Artifact` since this will allow use to have all the code plumbing done for all the builders for when we do implement it for them. I think `Jib` and `Kaniko` also support base images (maybe multistage also? don't know that) and it should be possible to implement this feature for them too.

Reply:
> Artifact dependencies apply to other builder types, and even between builder types too.

Agreed. The current design only supports implementing it for dockerfile builder by relying on the `ARG` command in the Dockerfile and passing currently built image to override the arg.
For other builders, jib, buildpacks, etc, design & implementation plan is needed.

The only reason, i prefer moving this to `DockerArtifact` is having at artifact level will allow users (especially ide users) to configure "artifact dependency tag" on *any* artifact type. 
Letting users configure this for any artifact type but spitting out an error later isn't very user friendly *especially* if there is a way to avoid it.


>I want to leave this inside Artifact since this will allow use to have all the code plumbing done for all the builders for when we do implement it for them. I think Jib and Kaniko also support base images (maybe multistage also? don't know that) and it should be possible to implement this feature for them too.

Agreed. We can add the field to individual types as we implement them and move it to global once all artifacts have the support.

Reply:
re
> I disagree with scoping this to docker-artifacts. Artifact dependencies apply to other builder types, and even between builder types too.

Do you mean we should address other builds in this DD as well ?

Reply:
as discussed, let's broaden the scope of this design to all builders. the implementation can start with docker artifacts since it's probably the most complicated.

Reply:
> @tejal29: We can add the field to individual types as we implement them and move it to global once all artifacts have the support.

💡!  Now I understand what you mean.  +1 to putting this on DockerArtifact.